enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,386 @@
+}
+
+// PolicyFulcioSubject defines the OIDC issuer and the email of the fulcio authentication configuration.
+type PolicyFulcioSubject struct {
+	// oidcIssuer contains the expected OIDC issuer.
+	// Example: "https://expected.OIDC.issuer/"
+	// Required if rootOfTrust.fulcioCAData is set.
+	// +required
+	OIDCIssuer string `json:"oidcIssuer,omitempty"`
+
+	// signedEmail holds the email address the the certificate is issued for.
+	// Example: "expected-signing-user@example.com"
+	// Required if rootOfTrust.fulcioCAData is set.
+	// +required
+	SignedEmail string `json:"signedEmail,omitempty"`
+}
+
+// PolicyIdentity defines image identity the signature claims about the image.
+type PolicyIdentity struct {
+	// matchPolicy sets the type of matching to be used.
+	// Valid values are "MatchRepoDigestOrExact", "MatchRepository", "ExactRepository", "RemapIdentity". When omitted, the default value is "MatchRepoDigestOrExact".
+	// If set matchPolicy to ExactRepository, then the repository must be specified.
+	// If set matchPolicy to RemapIdentity, then the remapPrefix, and remapSignedPrefix must be specified.
+	// "MatchRepoDigestOrExact" means that the identity in the signature must be in the same repository as the image identity if the image identity is referenced by a digest. Otherwise, the identity in the signature must be the same as the image identity.
+	// "MatchRepository" means that the identity in the signature must be in the same repository as the image identity.
+	// "ExactRepository" means that the identity in the signature must be in the same repository as a specific identity specified by "repository".
+	// "RemapIdentity" means that the signature must be in the same as the remapped image identity. Remapped image identity is obtained by replacing the "remapPrefix" with the specified “remapSignedPrefix” if the the image identity matches the specified remapPrefix.
+	// +kubebuilder:validation:Enum=MatchRepoDigestOrExact;MatchRepository;ExactRepository;RemapIdentity
+	// +required
+	MatchPolicy IdentityMatchPolicy `json:"matchPolicy,omitempty"`
+	// reference is the reference of the image identity to be matched. This field is required if identityMatchPolicy is set to "exactReference".
+	// +optional
+	Repository string `json:"repository,omitempty"`
+	// remapPrefix is the prefix of the image identity to be matched. This field is required if identityMatchPolicy is set to "RemapIdentity".
+	// +optional
+	RemapPrefix string `json:"remapPrefix,omitempty"`
+	// remapSignedPrefix is the prefix of the image identity to be matched in the signature. This field is required if identityMatchPolicy is set to "RemapIdentity".
+	// +optional
+	RemapSignedPrefix string `json:"remapSignedPrefix,omitempty"`
+}
+
+// IdentityMatchPolicy defines the type of matching for setting "matchPolicy".
+// Valid values are "MatchRepoDigestOrExact", "MatchRepository", "ExactRepository", "RemapIdentity". When omitted, the default value is "MatchRepoDigestOrExact".
```

Comment:
It is always guaranteed. If the field is not specified, the container runtime will always use the `MatchRepoDigestOrExact` as the default.

Reply:
Can the container runtime guarantee the default value will never change? Safer to state the default value is a subject to change. E.g.:
```
// When omitted, the container runtime defaults to "MatchRepoDigestOrExact".
// These defaults are subject to change over time.
```

Reply:
I’d say it’s… sufficiently guaranteed; it has been stable for longer than OCP 4 has existed:

The default is a documented property of the config file. That file is intended to be used in production deployments and we don’t intend to break users.

Uh… except that we did change the default, once. But that was in 2016, within the first 3 months of the existence of `policy.json` (and before Podman existed). And that change did not, AFAIK, break any existing deployments; it allowed new situations where a valid signature was accepted. That could have, in theory, broken someone’s security design, but that would have been a very unexpected security design.

Reply:
Pragmatically speaking, most users should not need to make a choice here, and for those users we would want to, similarly, reserve the _technical capability_ to change the behavior to “continue to do the right thing under changing circumstances” (which I hope we would never need to do).

I think anything that forces users to think what `repoDigestOrExact` means for them, or suggests to them that they should be explicitly hard-coding that policy, is a net detriment.

---

Also, there’s still one more layer of compatibility levers: Even if `policy.json` did change in incompatible ways, there’s still OCP’s `machine-config-operator` in between, so the MCO could adapt to that (hypothetical and expected) `policy.json` change by generating a `policy.json` that hard-codes the compatible choice.

IOW, OCP is fully in control of the compatibility here, so I see absolutely no reason to seed doubt about OCP’s _capability_ to not break users, nor to make anything more complex for users in order to preserve workloads.

---

That said, whether we document this “the default value is `MatchRepoDigestOrExact`”, or “the default value Does The Right Thing, don’t you worry”, I can see arguments for either.

Given the almost-7-hear-history of the current default, I think the former is preferable because it reassures and explains things better.